### PR TITLE
Adding inner window and multi window containers

### PR DESCRIPTION
**Pull request by andydotxyz**
### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.
- [x] Public APIs match existing style and have Since: line.

innerwin

### CODE REVIEW

**Comment by andydotxyz** [comment]
Feel free to test it out. Much testing and docs required I think, but looking for feedback on the API & widgets/layout etc

**Comment by andydotxyz** [comment]
Same windows build issues as https://github.com/fyne-io/fyne/pull/4262 - unrelated to this code

**Comment by c1ngular** [comment]
What is this for ? Window embedding ?

**Comment by andydotxyz** [comment]
> What is this for ? Window embedding ?

for making some content of your app appear within a window inside the app. Like the MDI interfaces from Windows https://en.wikipedia.org/wiki/Multiple-document_interface

**Comment by c1ngular** [comment]
@andydotxyz  i don't understand , so it is not an actual native window , but an UI widget/container looks like a window visually ?

**Comment by andydotxyz** [comment]
It looks like a window. Nothing to do with native windows in border or content. See screenshot at the top or try out the fyne_demo in this branch.

**Comment by c1ngular** [comment]
@andydotxyz  I see , thanks

File: container/innerwindow.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+func (i *innerWindowRenderer) Layout(size fyne.Size) {
+	pad := theme.Padding()
+	pos := fyne.NewSquareOffsetPos(pad / 2)
+	size = size.Subtract(fyne.NewSquareSize(pad))
+	i.LayoutShadow(size, pos)
+
+	i.bg.Move(pos)
+	i.bg.Resize(size)
+
+	pad = theme.Padding()
```
Comment (Jacalz) [inline container/innerwindow.go:109]: Are you not setting the value to the same one here?

**Review by Jacalz — CHANGES_REQUESTED** [review]
Nice work. This is very cool. I left a comment inline but also have some thoughts below. You don't necessarily have to address all of them but I think it might be wise to at least consider them.
- Resizing does not change the mouse cursor. Is that expected?
- I think the title would look better if it was centered in the window instead of being left-aligned.
- I feel like we need some sort of way to control the side of the buttons. It looks a bit out of place with inner windows having buttons to the left and my main window having them to the right. I also wander which of these should be the default? With Windows users being more than Mac-users I suspect that controls on the right side would be preferable?
-  The icon that is displayed in the top right is clickable but does nothing when it is clicked. Should it just be an icon and not a button?
- There seems to be less padding on the top of the window than on the other sides? See image below.

image
image

Sorry for the rambly sort of review. This really is great stuff but I think it can be made even better :)

**Comment by andydotxyz** [comment]
> * Resizing does not change the mouse cursor. Is that expected?

GLFW does not have a standard cursor for this yet sadly :( - in the future this could be addressed

> I feel like we need some sort of way to control the side of the buttons. It looks a bit out of place with inner windows having buttons to the left and my main window having them to the right. I also wander which of these should be the default? With Windows users being more than Mac-users I suspect that controls on the right side would be preferable? Or do we change default depending on the OS?

Let's address this in a future release and attempt to switch them based on OS. Not sure how we juggle GTK vs KDE vs Enlightenment though!

> There seems to be less padding on the top of the window than on the other sides? See image below.

The padding on 3 sides is the size of padding, on the top it is the window bar - adding more padding didn't look good

I have addressed the other points. If you agree with the summary above we should be good. I can add a ticket to add and detect the button side so it's not forgotten.

**Comment by Jacalz** [comment]
I think the centered label would have to be centered in the window, not centered between the buttons. Sorry. I understand that it makes the layout a bit more complicated but a relatively simple custom layout should do the job
image

**Comment by andydotxyz** [comment]
> I think the centered label would have to be centered in the window, not centered between the buttons. Sorry. I understand that it makes the layout a bit more complicated but a relatively simple custom layout should do the job
> 
> image
> 

If you want the window to be propped open by title that's a simple custom layout - but this really needs truncation.
What should we truncate to if it's centered in the window?

**Comment by Jacalz** [comment]
Why would having it cantered in the window break truncation? As long as the custom layout handles MinSize correctly, it seems like it should work. All my other windows on my computer has truncation for the title and test in the centre.

**Comment by andydotxyz** [comment]
Because there is a different space required on the left and right it cannot simply be centered with truncate turned on.
It would have to center as long as there is space and then switch to some other mode when the space needed is less than available if assuming even padding but more than is actually required because the right side is so much smaller in terms of button space.

That seems quite complex and given that left align is standard on Windows 11 I'm tempted to go back to that for now rather than chase this truncation algorithm...

**Review by Jacalz — APPROVED** [review]
Agreed. Approved

**Review by Jacalz — APPROVED** [review]

**Review by Jacalz — APPROVED** [review]

**Pull request merged**